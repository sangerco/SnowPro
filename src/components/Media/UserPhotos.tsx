import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchPhotosByUsername } from "../../redux/slices/mediaSlices";
import { useParams } from "react-router";
import { Button, Dimmer, Header, Loader } from "semantic-ui-react";
import { useNavigate } from "react-router";
import PhotoView from "./PhotoView";

const UserPhotos: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const { username } = useParams();

  useEffect(() => {
    if (username) {
      dispatch(fetchPhotosByUsername(username));
    }
  }, [dispatch, username]);

  const media = useSelector((state: RootState) => state.media);
  const photos = media.photos;

  if (media.loading) {
    <Dimmer active>
      <Loader>Loading...</Loader>
    </Dimmer>;
  }

  if (media.error) {
    <Dimmer active>
      <Header as="h1">Error! Pics could not be retrieved! {media.error}</Header>
      <Button color="red" onClick={() => navigate(-1)}>
        Go Back
      </Button>
    </Dimmer>;
  }

  if (photos) {
    return (
      <div>
        {photos.length > 0
          ? photos.map((photo) => <PhotoView key={photo.id} id={photo.id} />)
          : null}
      </div>
    );
  }

  return null;
};

export default UserPhotos;
