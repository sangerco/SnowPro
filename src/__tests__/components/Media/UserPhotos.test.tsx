/* eslint-disable testing-library/no-node-access */
import React from "react";
import UserPhotos from "../../../components/Media/UserPhotos";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test User Photos", () => {
  const store = mockStore({
    auth: {
      data: {
        id: "11",
      },
    },
    media: {
      photos: [
        {
          userId: "11",
          username: "testuser",
          id: "photo-1",
          link: "https://www.testlinks.com/image-1.jpg",
          about: "test about photo",
          createdAt: "fake created date",
        },
        {
          userId: "11",
          username: "testuser",
          id: "photo-2",
          link: "https://www.testlinks.com/image-2.jpg",
          about: "test about photo",
          createdAt: "fake created date",
        },
      ],
    },
  });
  it("renders without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPhotos />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPhotos />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("testing photo views", () => {
  it(`show a user's photos`, async () => {
    const username = "testuser";
    const useParamsMock = jest.requireMock("react-router-dom")
      .useParams as jest.Mock;
    useParamsMock.mockReturnValue({ username });

    const store = mockStore({
      auth: {
        data: {
          id: "11",
        },
      },
      media: {
        photos: [
          {
            userId: "11",
            username: "testuser",
            id: "photo-1",
            link: "https://www.testlinks.com/image-1.jpg",
            about: "test about photo",
            createdAt: "fake created date",
          },
          {
            userId: "11",
            username: "testuser",
            id: "photo-2",
            link: "https://www.testlinks.com/image-2.jpg",
            about: "test about photo",
            createdAt: "fake created date",
          },
        ],
      },
    });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserPhotos />
        </BrowserRouter>
      </Provider>
    );

    // @ts-ignore
    const photos = store.getState((state) => state.media.photos);
    // @ts-ignore
    expect(photos.media.photos.length).toBe(2);
  });
});
