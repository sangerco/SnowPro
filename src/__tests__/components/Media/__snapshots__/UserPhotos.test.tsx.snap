// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User Photos matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
